garage/tf/policies/gaussian_mlp_policy.py

```diff
@@ -47,214 +47,89 @@ def __init__(self,
-            else:
-                raise NotImplementedError
-
-            self.min_std_param = min_std_param
-
-            # mean_var, log_std_var = L.get_output([l_mean, l_std_param])
-            #
-            # if self.min_std_param is not None:
-            #     log_std_var = tf.maximum(log_std_var, np.log(min_std))
-            #
-            # self._mean_var, self._log_std_var = mean_var, log_std_var
-
-            self._l_mean = l_mean
-            self._l_std_param = l_std_param
-
-            self._dist = DiagonalGaussian(action_dim)
-
-            LayersPowered.__init__(self, [l_mean, l_std_param])
-            super(GaussianMLPPolicy, self).__init__(env_spec)
-
-            dist_info_sym = self.dist_info_sym(
-                mean_network.input_layer.input_var, dict())
-            mean_var = tf.identity(dist_info_sym["mean"], name="mean")
-            log_std_var = tf.identity(
-                dist_info_sym["log_std"], name="standard_dev")
-
-            self._f_dist = tensor_utils.compile_function(
-                inputs=[obs_var],
-                outputs=[mean_var, log_std_var],
-            )
+        
+        self.model = GaussianMLPModel(
+            input_dim=env_spec.observation_space.flat_dim,
+            output_dim=env_spec.action_space.flat_dim,
+            hidden_sizes=hidden_sizes,
+            learn_std=learn_std,
+            init_std=init_std,
+            adaptive_std=adaptive_std,
+            std_share_network=std_share_network,
+            std_hidden_sizes=std_hidden_sizes,
+            min_std=min_std,
+            max_std=max_std,
+            std_hidden_nonlinearity=std_hidden_nonlinearity,
+            hidden_nonlinearity=hidden_nonlinearity,
+            output_nonlinearity=output_nonlinearity,
+            std_parameterization=std_parameterization,
+        )
+
+        self._dist = DiagonalGaussian(self.env_spec.action_space.flat_dim)
 
     @property
     def vectorized(self):
         return True
 
-    def dist_info_sym(self, obs_var, state_info_vars=None, name=None):
-        with tf.name_scope(name, "dist_info_sym", [obs_var]):
-            with tf.name_scope(self._mean_network_name, values=[obs_var]):
-                mean_var = L.get_output(self._l_mean, obs_var)
-            with tf.name_scope(self._std_network_name, values=[obs_var]):
-                std_param_var = L.get_output(self._l_std_param, obs_var)
-            if self.min_std_param is not None:
-                std_param_var = tf.maximum(std_param_var, self.min_std_param)
-            if self.std_parametrization == 'exp':
-                log_std_var = std_param_var
-            elif self.std_parametrization == 'softplus':
-                log_std_var = tf.log(tf.log(1. + tf.exp(std_param_var)))
-            else:
-                raise NotImplementedError
-            return dict(mean=mean_var, log_std=log_std_var)
-
-    @overrides
     def get_action(self, observation):
+        sess = tf.get_default_session()
```

Comment:
This really makes me nervous.

Shouldn't we have a global session singleton if we're going to do this?

Reply:
This imply the default session of the thread is used but we can always have the session as argument.

Reply:
I know. I am positing we should eliminate this assumption in the TF tree, since the session can change at any time anywhere in the code.

That's probably out-of-scope for this PR.

However, you should at lease provide `sess=get_default_session()` as an optional kwarg, so that callers can control which session you're using.